src/main/drivers/flash_w25n.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * This file is part of Cleanflight and Betaflight.
+ *
+ * Cleanflight and Betaflight are free software. You can redistribute
+ * this software and/or modify this software under the terms of the
+ * GNU General Public License as published by the Free Software
+ * Foundation, either version 3 of the License, or (at your option)
+ * any later version.
+ *
+ * Cleanflight and Betaflight are distributed in the hope that they
+ * will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "platform.h"
+
+#include "build/debug.h"
+
+#ifdef USE_FLASH_W25N
+
+#include "common/maths.h"
+#include "drivers/bus_spi.h"
+#include "drivers/flash.h"
+#include "drivers/flash_impl.h"
+#include "drivers/io.h"
+#include "drivers/time.h"
+
+#include "pg/flash.h"
+
+#include "flash_w25n.h"
+
+#define W25N_INSTRUCTION_RDID                       SPIFLASH_INSTRUCTION_RDID
+#define W25N_INSTRUCTION_PAGE_DATA_READ             0x13
+#define W25N_INSTRUCTION_READ_DATA                  0x03
+#define W25N_INSTRUCTION_READ_STATUS_REG            0x05
+#define W25N_INSTRUCTION_WRITE_STATUS_REG           0x01
+#define W25N_INSTRUCTION_WRITE_ENABLE               0x06
+#define W25N_INSTRUCTION_WRITE_DISABLE              0x04
+#define W25N_INSTRUCTION_LOAD_PROGRAM_DATA          0x02
+#define W25N_INSTRUCTION__RANDOMLOAD_PROGRAM_DATA   0x84
+#define W25N_INSTRUCTION_PROGRAM_EXECUTE            0x10
+#define W25N_INSTRUCTION_BLOCK_ERASE                0xD8
+
+#define W25N_STATUS_Register_3                      0xC0
+
+#define W25N_STATUS_FLAG_WRITE_ENABLE_LATCH         0x01
+#define W25N_STATUS_FLAG_BUSY                       0x02
+
+// Format is manufacturer, memory type, then capacity
+#define JEDEC_ID_WINBOND_W25N01GV                   0xEFAA21
+
+// The timeout we expect between being able to issue page program instructions
+#define DEFAULT_TIMEOUT_MILLIS                      6
+
+#define READ_PAGE_DATA_TIMEOUT_MILLIS               60
+#define PAGE_PROGRAM_TIMEOUT_MILLIS                 700
+#define PROGRAM_EXECUTE_TIMEOUT_MILLIS              10
+#define BLOCK_ERASE_TIMEOUT_MILLIS                  10000
+
+#define W25N_PAGESIZE                               2048
+
+STATIC_ASSERT(W25N_PAGESIZE <= FLASH_MAX_PAGE_SIZE, W25N_PAGESIZE_too_small);
+
+const flashVTable_t w25n_vTable;
+
+static void w25n_disable(busDevice_t *bus)
+{
+    IOHi(bus->busdev_u.spi.csnPin);
+    __NOP();
```

Comment:
I think you want a barrier here, not a `nop`.

Reply:
This part was taken from flash_m25p16.c. So we may also review it there?

Reply:
Sure, we should